File: 2015/Day7/main.lua
local wire_values = {}
local instructions = {}

local function get_wire_value(wire)
    if wire_values[wire] then
        return wire_values[wire]
    end
    if tonumber(wire) then
        return tonumber(wire)
    end

    local instruction = instructions[wire]
    local result

    if #instruction == 1 then
        result = get_wire_value(instruction[1])
    elseif #instruction == 2 then
        result = ~get_wire_value(instruction[2])
    else
        local left = get_wire_value(instruction[1])
        local right = get_wire_value(instruction[3])
        local op = instruction[2]

        if op == "AND" then
            result = left & right
        elseif op == "OR" then
            result = left | right
        elseif op == "LSHIFT" then
            result = left << right
        elseif op == "RSHIFT" then
            result = left >> right
        else
            error("Unknown operator: " .. op)
        end
    end

    wire_values[wire] = result
    return result
end

local function solve_circuit(override_b, b_value)
    wire_values = {}
    if override_b then
        wire_values["b"] = b_value
    end
    return get_wire_value("a")
end

for line in io.lines('input.txt') do
    local tokens = {}
    for token in string.gmatch(line, "%S+") do
        if token ~= "->" then
            table.insert(tokens, token)
        end
    end
    local wire = table.remove(tokens)
    instructions[wire] = tokens
end

local part1 = get_wire_value("a")
print("Part 1: " .. part1)
print("Part 2: " .. solve_circuit(true, part1))

File: 2015/Day7/main.dart
import 'dart:io';

Map<String, int> wireValues = {};
Map<String, List<String>> instructions = {};

int getWireValue(String wire) {
  if (wireValues.containsKey(wire)) {
    return wireValues[wire]!;
  }
  if (int.tryParse(wire) != null) {
    return int.parse(wire);
  }

  List<String> instruction = instructions[wire]!;
  late int result;

  if (instruction.length == 1) {
    result = getWireValue(instruction[0]);
  } else if (instruction.length == 2) {
    result = ~getWireValue(instruction[1]) & 0xffff;
  } else {
    int left = getWireValue(instruction[0]);
    int right = getWireValue(instruction[2]);
    String op = instruction[1];

    if (op == 'AND') {
      result = left & right;
    } else if (op == 'OR') {
      result = left | right;
    } else if (op == 'LSHIFT') {
      result = left << right;
    } else if (op == 'RSHIFT') {
      result = left >> right;
    } else {
      throw Exception('Unknown operator: $op');
    }
  }

  wireValues[wire] = result;
  return result;
}

int solveCircuit({bool overrideB = false, int bValue = 0}) {
  wireValues.clear();
  if (overrideB) {
    wireValues['b'] = bValue;
  }
  return getWireValue('a');
}

void main() async {
  final lines = await File('input.txt').readAsLines();

  for (var line in lines) {
    List<String> tokens = line.split(' ');
    String wire = tokens.last;
    tokens.removeLast();
    tokens.removeLast(); // Remove the '->'
    instructions[wire] = tokens;
  }

  int part1 = getWireValue('a');
  print('Part 1: $part1');
  print('Part 2: ${solveCircuit(overrideB: true, bValue: part1)}');
}
